### ISSUE TITLE
create a mapping between AST's and syntactical/logical skills

### ISSUE THREAD

**Issue by diraneyya** [diraneyya/js-huggle#2]
This is a theoretical step that comes in preparation for my new skills map architecture.

I want to create a systematic method for converting the AST of the desired outcome of a level to the skills this level reinforces.

The value of the skill will be accumulated via repetition and will follow a memory curve to emphasise the importance of timed repetition in learning.

But for the time being some form of theoretical planning is needed.

Outcome:
- Choose an off-the-shelf AST generator
- Create an algorithmic mapping between the AST of the solution of a level and the skills object.

This, in itself, sets JS-huggle to be a syntactical concept. The idea is that the block in learning a programming language for the first time does not come from the semantic and the advanced understanding of the use cases of the language, rather, it results from the huge leap in thinking to adapt to the _syntax of logical languages_, which sets them apart from _natural languages_.

Namely, the astounding consistency and logical conformance of a logical language. This usually is met with a sloppy and an imprecise mindset at those with life adaptations for mastering natural languages. Mostly hustlers, house wives, hands-on workers, social workers, HR people, etc.

**Comment by diraneyya** [comment]
## TL;DR

We want to map solutions to ASTs and then to syntactical skills.

For example:
- "object.accessMe" might map to skills such as "access property of an object"
- "object.games[index].map(...)" might map to skills such as:
    - access property of an object
    - index an array with an identifier
    - map an array element using ...
    
And so on...

**Comment by diraneyya** [comment]
Once I build this mapping I can start creating the "level mixer" which is an engine that using your current skills will create a level just for you.

Like if you are learning German, you would want someone to tell you a sentence in which you know all the words except one word, for example.

moderate challenge => maximum learning

Hence levels increase and decrease in complexity automatically. If you struggle they go less complex, if you do well, they go more complex and start incorporating syntactical features that are dependent on the syntactical features you mastered.
